### PR TITLE
Fix stack pointer retrieval in jl_backtrace_from_here

### PR DESCRIPTION
**Pull request by tkf**
It seems `jl_backtrace_from_here` with the `returnsp` flag set was broken. This patch adds a few fixes and a test.

### CODE REVIEW

**Comment by tkf** [comment]
Hmm... The failure in win32 seems to be a real one: https://build.julialang.org/#/builders/9/builds/4162/steps/5/logs/stdio

```
Test Failed at C:\buildbot\worker-tabularasa\tester_win32\build\share\julia\test\backtrace.jl:331
  Expression: ptr2 < sp[1] < ptr1
   Evaluated: Ptr{Nothing} @0x2a07e1b8 < Ptr{Nothing} @0x2a07e1b0 < Ptr{Nothing} @0x2a07e228
```

i.e., `ptr2 < sp[1]` doesn't hold because `0x2a07e1b8 - 0x2a07e1b0 == 8 > 0`. But I have no idea why...

File: test/backtrace.jl [hunk 1]
```diff
@@ -258,3 +258,76 @@ let code = """
...
+end
+
+"""
+    withframeaddress(f)
+
+Call function `f` with an address `ptr::Ptr{Cvoid}` of an independent frame
+immediately outer to `f`.
+"""
+@noinline function withframeaddress(f)
+    sp = ccall("llvm.frameaddress", llvmcall, Ptr{Cvoid}, (Int32,), 0)
```
Comment (vtjnash) [inline test/backtrace.jl:312]: the return type here is wrong (it should be `i8*`)

**Review by vtjnash — COMMENTED** [review]
The code for win32 looks reasonable:
```julia
# wine ./julia -e "@noinline function withframeaddress()
#            sp = ccall(\"llvm.frameaddress\", llvmcall, Int, (Int32,), 0)
#            (@noinline function(sp) sp end)(sp)
# end
# using InteractiveUtils; @code_native withframeaddress()"

        .text
; ┌ @ none:1 within `withframeaddress'
        pushl   %ebp
        movl    %esp, %ebp
        andl    $-16, %esp
        subl    $16, %esp
; │ @ none:3 within `withframeaddress'
        movl    %ebp, (%esp)
        calll   "#1"
        movl    %ebp, %esp
        popl    %ebp
        retl
        nopw    %cs:(%eax,%eax)
        nop
; └
```

So the trouble is perhaps that `jl_backtrace_from_here` is defined to push 8 bytes on the stack (changing SP) first (for true::Cint and 0::Cint arguments) on x86, and then has the additional realignment calls too for win32

**Review by vtjnash — COMMENTED** [review]
Actually, I don't understand this stack frame, though you can see here how $esp@ jl_backtrace_from_here calll > $esp@withframeaddress entry (by 8 bytes):
```
# wine ./julia -e "@noinline function withframeaddress(g)
#            sp = ccall(\"llvm.frameaddress\", llvmcall, Int, (Int32,), 0); g(sp) end
#using InteractiveUtils
#code_native(()) do
#         bt = ccall(:jl_backtrace_from_here, Ref{Base.SimpleVector}, (Cint, Cint), true, 0)
#         withframeaddress(@noinline g(p2) = p2)
#end"
        .text
; ┌ @ none:5 within `#1'
        pushl   %ebp
        movl    %esp, %ebp
        andl    $-16, %esp
        subl    $24, %esp
        movl    $jl_backtrace_from_here, %eax
        pushl   $0
        pushl   $1
        calll   *%eax
        addl    $16, %esp
; │ @ none:6 within `#1'
        calll   withframeaddress
        movl    %ebp, %esp
        popl    %ebp
        retl
; └
```

**Comment by vtjnash** [comment]
(IIRC, the past version of the code may have been attempting to account for this by using the stackpointer prior to the call as the value for that frame, instead of using the value for the current frame, but since almost nobody used the value, it was likely to be offset even more for many edge cases)

File: test/backtrace.jl [hunk 2]
```diff
@@ -258,3 +258,76 @@ let code = """
...
+            ptr2 = p2
+        end
+    end
+    return ptr1, ptr2, bt
+end
+
+@testset "stack pointers" begin
+    ptr1, ptr2, bt_data = sandwiched_backtrace()
+    sp = _reformat_sp(bt_data...)
+    @test ptr2 < sp[1] < ptr1
```
Comment (vtjnash) [inline test/backtrace.jl:331]:
```suggestion
    @test ptr2 < sp[2] < ptr1
```

File: test/backtrace.jl (hunk 1 above)
Comment (vtjnash) [reply test/backtrace.jl:312]: Probably need to do this instead:
```julia
julia> @eval frameaddr() = Core.Intrinsics.llvmcall(($"""
       declare i8* @llvm.frameaddress(i32)
       define i$(Sys.WORD_SIZE) @frameaddr() {
         %1 = call i8* @llvm.frameaddress(i32 0)
         %2 = ptrtoint i8* %1 to i$(Sys.WORD_SIZE)
         ret i$(Sys.WORD_SIZE) %2
       }""", "frameaddr"), UInt, Tuple{})
frameaddr (generic function with 1 method)

julia> frameaddr()
0x00007ffdac2bd2d0    
```

Comment (tkf) [reply test/backtrace.jl]: added 715c673a889a2c8a9b81b5628db9e383ab47b8c9

File: test/backtrace.jl [hunk 3]
```diff
@@ -258,3 +258,91 @@ let code = """
...
+            j += 1
+            if bt_raw[j] == ip
+                push!(sp, sp_raw[j])
+                break
+            end
+        end
+    end
+end
+
+@eval @inline function frameaddr()
```
Comment (vtjnash) [inline test/backtrace.jl:305]: I think you can literally inline this into `withframeaddress`. I just posted it separately to simplify the comment

**Comment by tkf** [comment]
@vtjnash Don't we need something like 3f8868a181b06258e7fba99eec7814121129caa2 to make it work on all platforms?

I guess the innermost `noinlinecall` is redundant though.

**Comment by tkf** [comment]
The failures in FreeBSD seem irrelevant https://build.julialang.org/#/builders/78/builds/5229. It passes the backtrace test.

**Comment by vtjnash** [comment]
I thought https://github.com/JuliaLang/julia/commit/3f8868a181b06258e7fba99eec7814121129caa2 was simply needed if we want to ensure we the test would pass even if we were getting the wrong answer from the intermediate functions. Though either answer seems almost correct for what is expected of the stack pointer.

**Comment by tkf** [comment]
I was getting the failure in all platforms without 3f8868a181b06258e7fba99eec7814121129caa2.

In linux64 https://build.julialang.org/#/builders/69/builds/6083/steps/5/logs/stdio

```
Test Failed at /buildworker/worker/tester_linux64/build/share/julia/test/backtrace.jl:342
  Expression: ptr2 < sp[2] < ptr1
   Evaluated: Ptr{Nothing} @0x00007f6991e99f80 < Ptr{Nothing} @0x00007f6991e99ff0 < Ptr{Nothing} @0x00007f6991e99fe0
```

In win64 https://build.julialang.org/#/builders/65/builds/5392

```
Test Failed at C:\buildbot\worker-tabularasa\tester_win64\build\share\julia\test\backtrace.jl:342
  Expression: ptr2 < sp[2] < ptr1
   Evaluated: Ptr{Nothing} @0x0000000063c6cb20 < Ptr{Nothing} @0x0000000063c6cbe0 < Ptr{Nothing} @0x0000000063c6cbd0
```

In win32 https://build.julialang.org/#/builders/9/builds/5488

```
Test Failed at C:\buildbot\worker-tabularasa\tester_win32\build\share\julia\test\backtrace.jl:342
  Expression: ptr2 < sp[2] < ptr1
   Evaluated: Ptr{Nothing} @0x6b57e228 < Ptr{Nothing} @0x6b57e2a0 < Ptr{Nothing} @0x6b57e298
```

In these cases, `ptr2 < sp[2]` holds but `sp[2] < ptr1` does not.

Since `jl_backtrace_from_here` seems to try to skip its own frame, I thought I'd makes sense (at least in non-Windows?) to insert the outer `noinlinecall` since `sp[2]` will be the frame in which we call `@llvm.frameaddress`. (Though I don't understand why we see `sp[2] > ptr1`... Is it possible that the definition of frameaddress is different in LLVM and libunwind...?)

Is it OK to pull this as-is or do you want to investigate this further?

**Comment by vtjnash** [comment]
Perhaps it would be `ptr2 < sp[2] && sp[1] < ptr1`? I am not sure, and am okay with this as an improvement either way

**Comment by tkf** [comment]
Ah, good point. Let's see if it works.

**Comment by tkf** [comment]
Hmm... win64 failed but I don't think if it's relevant https://build.julialang.org/#/builders/65/builds/5439 (but re-starting it to be sure)

FreeBSD failure seems irrelevant.

**Comment by DilumAluthge** [comment]
If you rebase this PR branch on the latest master, that will fix the FreeBSD test failures.

Also, I noticed that the master branch of your fork is behind the master branch of JuliaLang/julia. I'd recommend fast-forwarding the master branch of your fork so that it is up-to-date with the master branch of JuliaLang/julia.

**Comment by tkf** [comment]
Thanks, I didn't realize that this branch is rather behind the `master`. I just merged it.

I always start a branch from whatever the `master` of JuliaLang/julia is. So I don't think `master` of my GitHub fork is relevant here?

**Comment by DilumAluthge** [comment]
> I always start a branch from whatever the `master` of JuliaLang/julia is. So I don't think `master` of my GitHub fork is relevant here?

Ah, I thought you were starting feature branches off of the `master` of your fork. If not, yeah the `master` of your fork is not relevant.

**Comment by KristofferC** [comment]
@tkf, the tests on https://github.com/JuliaLang/julia/pull/43084 fails due to this with `  LoadError: f(Ptr{Cvoid}(sp)) is not a function expression`. Due you have a version of the tests which is backportable?

**Pull request merged**